In [3]:
import numpy as np

def print_matrix(matrix, label):
    print(label)
    for row in matrix:
        print(row)

def qr_decomposition(A):
    m, n = A.shape
    Q = np.zeros((m, n))
    R = np.zeros((n, n))

    for j in range(n):
        v = A[:, j]

        for i in range(j):
            R[i, j] = np.dot(Q[:, i], A[:, j])
            v = v - R[i, j] * Q[:, i]

        R[j, j] = np.linalg.norm(v)
        Q[:, j] = v / R[j, j]

    return Q, R

# Generate random matrix A
# np.random.seed(42)  # Set seed for reproducibility
m = 4
n = 3
A = np.random.randint(1, 10, size=(m, n)).astype(float)

print_matrix(A, "Original Matrix A:")

# Perform QR decomposition
Q, R = qr_decomposition(A)

print_matrix(Q, "Matrix Q (Orthogonal):")
print_matrix(R, "Matrix R (Upper Triangular):")

# Verify QR = A with a tolerance
tolerance = 1e-8
verification_result = np.allclose(np.dot(Q, R), A, atol=tolerance)
print("\nVerification Result (QR = A):", verification_result)


Original Matrix A:
[8. 5. 4.]
[3. 4. 6.]
[9. 2. 1.]
[9. 8. 9.]
Matrix Q (Orthogonal):
[ 0.52186246  0.0344582  -0.63683953]
[0.19569842 0.4541414  0.6735758 ]
[ 0.58709527 -0.71390322  0.37375668]
[ 0.58709527  0.53189324 -0.03220236]
Matrix R (Upper Triangular):
[15.32970972  9.26305864  9.13259302]
[0.         4.81619608 6.93581714]
[0.         0.         1.57803209]

Verification Result (QR = A): True


Amit bhai 

Original Matrix A:
[8. 2. 7.]
[5. 9. 8.]
[1. 4. 4.]
[4. 7. 1.]
Matrix Q (Orthogonal):
[ 0.77702869 -0.60681981  0.09386099]
[0.48564293 0.55777234 0.39253814]
[0.09712859 0.37755142 0.45697661]
[ 0.38851434  0.42203633 -0.79263882]
Matrix R (Upper Triangular):
[10.29563014  9.03295852 10.10137297]
[0.         8.27077145 2.14668209]
[0.         0.         4.83259973]

Verification Result (QR = A): True

anshuman 

Original Matrix A:
[8. 5. 4.]
[3. 4. 6.]
[9. 2. 1.]
[9. 8. 9.]
Matrix Q (Orthogonal):
[ 0.52186246  0.0344582  -0.63683953]
[0.19569842 0.4541414  0.6735758 ]
[ 0.58709527 -0.71390322  0.37375668]
[ 0.58709527  0.53189324 -0.03220236]
Matrix R (Upper Triangular):
[15.32970972  9.26305864  9.13259302]
[0.         4.81619608 6.93581714]
[0.         0.         1.57803209]

Verification Result (QR = A): True

Vishal bhai


Original Matrix A:
[1. 6. 3.]
[6. 8. 1.]
[8. 8. 5.]
[4. 5. 6.]
Matrix Q (Orthogonal):
[0.09245003 0.94121259 0.11955963]
[ 0.5547002   0.1802322  -0.66345106]
[ 0.73960026 -0.2803612   0.11809135]
[0.36980013 0.05507095 0.72910398]
Matrix R (Upper Triangular):
[10.81665383 12.75810451  6.74885239]
[0.         5.12159831 1.93248968]
[0.         0.         4.66030847]

Verification Result (QR = A): True